### PR TITLE
Shop

### PR DESCRIPTION
**Pull request by bethios**
http://sacred-art-development.herokuapp.com/shop

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -86,10 +86,9 @@ nav{
 
 .ion-navicon-round{
   float:right;
-  margin-top: 20px;
-  margin-right: 5vw;
+  margin-top: 10px;
+  margin-right: 15px;
```
Comment (cblant20) [inline app/assets/stylesheets/application.scss:90]: I thought 15 was your standard size?

File: app/assets/stylesheets/application.scss [hunk 2]
```diff
@@ -124,9 +123,23 @@ nav{
   padding: 10px 0;
 }
 
-#dropdown-mobile{
-  right: 0;
-  font-size: 0.6em;
+.mobile-menu{
+  background: $offwhite;
```
Comment (cblant20) [inline app/assets/stylesheets/application.scss:127]: good

File: app/assets/stylesheets/application.scss [hunk 3]
```diff
@@ -206,6 +219,38 @@ div.footer{
...
+.btn-success{
+  background-color: $slate;
+}
+
+.btn-success:hover{
+  background-color: $background;
+  color: black;
+}
+
+.admin-buttons{
```
Comment (cblant20) [inline app/assets/stylesheets/application.scss:235]: non-standard naming; some are buttons while some are btn

Comment (cblant20) [reply app/assets/stylesheets/application.scss:235]: contextually, you would know it was a button if you combined class and element selectors

File: app/assets/stylesheets/events.scss [hunk 4]
```diff
@@ -0,0 +1,24 @@
+table{
+  width: 90%;
+  margin: 20px;
+  background-color: white;
+}
+
+td, th{
+  text-align: left;
+  padding: 15px;
+  font-size: 120%;
```
Comment (cblant20) [inline app/assets/stylesheets/events.scss:10]: remember font sizes should be in variables. Also keep your units consistent.

File: app/assets/stylesheets/events.scss [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+th {
+  font-size: 150%;
+  font-weight: bold;
+  border-bottom: 2px black solid;
+}
+tr:nth-child(even){
```
Comment (cblant20) [inline app/assets/stylesheets/events.scss:18]: missing a line break between these

File: app/assets/stylesheets/events.scss [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+td, th{
+  text-align: left;
+  padding: 15px;
+  font-size: 120%;
+}
+
+th {
+  font-size: 150%;
+  font-weight: bold;
+  border-bottom: 2px black solid;
```
Comment (cblant20) [inline app/assets/stylesheets/events.scss:16]: color variable!

File: app/assets/stylesheets/events.scss [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+}
+tr:nth-child(even){
+  background-color: #dddddd;
+}
+
+tr:nth-child(odd){
+  background-color: white;
```
Comment (cblant20) [inline app/assets/stylesheets/events.scss:23]: color variable!

File: app/assets/stylesheets/events.scss [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+}
```
Comment (cblant20) [inline app/assets/stylesheets/events.scss:24]: newline

File: app/assets/stylesheets/variables.scss [hunk 9]
```diff
@@ -5,7 +5,7 @@ $offwhite: rgba(234, 242, 239, 1);
 $accent: rgba(255, 102, 51, 1);
 $white: rgba(255,255,255,1);
```
Comment (cblant20) [inline app/assets/stylesheets/variables.scss:6]: see? you had it as a variable

File: app/controllers/faqs_controller.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+  def create
+    @faq = Faq.new
+    @faq.question = params[:faq][:question]
+    @faq.answer = params[:faq][:answer]
+
+    if @faq.save
+      redirect_to admin_path
+    else
+      flash[:alert] = 'error saving faq'
```
Comment (cblant20) [inline app/controllers/faqs_controller.rb:17]: fix your capitalization here

File: app/controllers/faqs_controller.rb [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+  def update
+    @faq = Faq.find(params[:id])
+    @faq.question = params[:faq][:question]
+    @faq.answer = params[:faq][:answer]
+
+    if @faq.save
+      redirect_to faqs_path
+    else
+      flash[:alert] = 'error saving faq'
```
Comment (cblant20) [inline app/controllers/faqs_controller.rb:29]: repeated code

File: app/controllers/hours_controller.rb [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+  before_action :authorize_user, except: :show
+
+  def index
+    @hours = Hour.all
+  end
+
+  def show
+    @hour = Hour.last
+    @hour.title.upcase!
+
```
Comment (cblant20) [inline app/controllers/hours_controller.rb:12]: needless newline

File: app/controllers/hours_controller.rb [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  def create
+    @hour = Hour.new
+    @hour.title = params[:hour][:title]
+    @hour.body = params[:hour][:body]
+
+    if @hour.save
+      redirect_to admin_path
+    else
+      flash[:alert] = 'error saving hours'
```
Comment (cblant20) [inline app/controllers/hours_controller.rb:27]: fix capitalization

File: app/controllers/hours_controller.rb [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+  def update
+    @hour = Hour.last
+    @hour.title = params[:hour][:title]
+    @hour.body = params[:hour][:body]
+
+    if @hour.save
+      redirect_to admin_path
+    else
+      flash[:alert] = 'error saving hours'
```
Comment (cblant20) [inline app/controllers/hours_controller.rb:43]: repeated code

File: app/controllers/hours_controller.rb [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    end
+  end
+
+  private
+
+  def authorize_user
```
Comment (cblant20) [inline app/controllers/hours_controller.rb:49]: this is private to the file, but I don't see you calling it?

File: app/controllers/welcome_controller.rb [hunk 16]
```diff
@@ -14,13 +17,22 @@ def new
...
-  end
-
   def submissions
   end
 
   def contact
   end
 
+  def admin
+
```
Comment (cblant20) [inline app/controllers/welcome_controller.rb:27]: all of these are blank?

File: app/controllers/welcome_controller.rb [hunk 17]
```diff
@@ -14,13 +17,22 @@ def new
...
+  end
+
+  private
+
+  def authorize_user
```
Comment (cblant20) [inline app/controllers/welcome_controller.rb:32]: this whole controller seems strange

File: app/models/artist.rb [hunk 18]
```diff
@@ -19,6 +22,7 @@ class Artist < ActiveRecord::Base
   validates_attachment_content_type :image_2, :content_type => /\Aimage\/.*\z/
 
   has_attached_file :image_3,
+                    :default_url => 'sa_initials.jpg',
```
Comment (cblant20) [inline app/models/artist.rb:25]: I am glad you handled the missing image case; however, what if you handled it by not displaying images at all there?

Comment (cblant20) [reply app/models/artist.rb:25]: Is there a way you could do a repeater for these 3 images since they are all very similar?

File: app/views/events/_current.html.erb [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+            <p><%= markdown(event.body) %></p>
+
+            <% if current_user && current_user.role == 'admin' %>
+              <%= link_to "Edit Event", edit_event_path(event.id), class: 'btn btn-success'  %>
+              <%= link_to "Delete Event", event, method: :delete, :class => 'btn btn-success' %>
+            <% end %>
+          </div>
+      <% end %>
+  <% end %>
+<% end %>
```
Comment (cblant20) [inline app/views/events/_current.html.erb:16]: newline

File: app/views/events/_form.html.erb [hunk 20]
```diff
@@ -0,0 +1,63 @@
+<div class="row">
+  <div class="col-sm-6">
+    <table class="content-container">
+      <tr>
+        <th>Style</th>
+        <th>Markdown Syntax</th>
+      </tr>
+      <tr>
+        <td>Autolink</td>
+        <td>< ></td>
```
Comment (cblant20) [inline app/views/events/_form.html.erb:10]: what is the open element tag?

File: app/views/events/_form.html.erb [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+            <%= f.text_area :body, rows: 12, col: 40, class: 'form-control' %>
+          </div>
+          <div class="form-group">
+            <%= f.submit "Save Event", class: 'btn btn-success'%>
+          </div>
+      <% end %>
+    </div>
+  </div>
+</div>
+
```
Comment (cblant20) [inline app/views/events/_form.html.erb:62]: extra line here

File: app/views/events/destroy.html.erb [hunk 22]
```diff
@@ -0,0 +1,2 @@
+<h1>Event#destroy</h1>
+<p>Find me in app/views/event/destroy.html.erb</p>
```
Comment (cblant20) [inline app/views/events/destroy.html.erb:2]: o,o

File: app/views/events/edit.html.erb [hunk 23]
```diff
@@ -0,0 +1,2 @@
+<h1 class="category-banner">Edit Event</h1>
+<%= render partial: 'form', locals: { event: @event }  %>
```
Comment (cblant20) [inline app/views/events/edit.html.erb:2]: newline

File: app/views/faqs/_form.html.erb [hunk 24]
```diff
@@ -0,0 +1,57 @@
+<div class="row">
+<div class="col-sm-6">
+  <table class="content-container">
```
Comment (cblant20) [inline app/views/faqs/_form.html.erb:3]: is this the same as in the form partial? can you share it between them?

File: app/views/faqs/_form.html.erb [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+          <%= f.label :answer %>
+          <%= f.text_area :answer, rows: 15, class: "form-control"  %>
+        </div>
+        <div class="form-group">
+          <%= f.submit "Save FAQ"  %>
+        </div>
+    <% end %>
+  </div>
+</div>
+</div>
```
Comment (cblant20) [inline app/views/faqs/_form.html.erb:57]: newline

File: app/views/faqs/index.html.erb [hunk 26]
```diff
@@ -0,0 +1,38 @@
+<div class="row">
+  <div class="col-sm-3 side-images">
+    <%= image_tag('boats.jpg', alt: 'small wooden boxes with tiny boats and waves inside that are controlled by a crank', class: 'content-container side-image') %>
+    <%= image_tag('bottles.jpg', alt: 'tiny rows of bottle stoppers with glass vessels on top with natural items inside', class: 'content-container side-image')%>
```
Comment (cblant20) [inline app/views/faqs/index.html.erb:4]: I am not sure what the image is from this alt text.

File: app/views/faqs/index.html.erb [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+          <div class="panel-heading">
+            <h4 class="panel-title">
+              <a data-toggle="collapse" href="#faq<%= faq.id %>" ><%= faq.question %> <span class="caret"></span></a>
+            </h4>
+          </div>
+          <div id="faq<%= faq.id %>"  class="panel-collapse collapse">
+            <div class="panel-body"><%= markdown(faq.answer) %>
+              <% if current_user && current_user.role == 'admin' %>
+                  <div class="update-buttons">
+                    <%= link_to "Edit FAQ", edit_faq_path(faq.id), :class => 'btn btn-success' %>
```
Comment (cblant20) [inline app/views/faqs/index.html.erb:26]: why are you doing  a line by line edit and not a big block edit?

File: app/views/faqs/index.html.erb [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+                    <%= link_to "Delete FAQ", faq, method: :delete, :class => 'btn btn-success' %>
+                  </div>
+              <% end %>
+            </div>
+          </div>
+        </div>
+      <% end %>
+
```
Comment (cblant20) [inline app/views/faqs/index.html.erb:34]: no need for this line

File: app/views/faqs/new.html.erb [hunk 29]
```diff
@@ -0,0 +1,2 @@
+<h2 class="category-banner">New FAQ</h2>
+<%= render partial: 'form', locals: { faq: @faq } %>
```
Comment (cblant20) [inline app/views/faqs/new.html.erb:2]: you know I want a newline

File: app/views/hours/_form.html.erb [hunk 30]
```diff
@@ -0,0 +1,24 @@
...
+          </div>
+          <div class="form-group">
+            <%= f.submit "Save" %>
+          </div>
+      <% end %>
+    </div>
+  </div>
+  <div class="col-sm-6">
+    <div class="content-container">
+      <p>Please keep to the current format for displaying store hours. It will format the text for you.</p>
```
Comment (cblant20) [inline app/views/hours/_form.html.erb:21]: I am not sure this is clear. Is there a way you can force them to?

File: app/views/hours/edit.html.erb [hunk 31]
```diff
@@ -0,0 +1,2 @@
+<h1 class="category-banner">Edit Hours</h1>
+<%= render partial: 'form', locals: { hours: @hours } %>
```
Comment (cblant20) [inline app/views/hours/edit.html.erb:2]: newline lala

File: app/views/hours/new.html.erb [hunk 32]
```diff
@@ -0,0 +1,2 @@
+<h1 class="category-banner">Add New Hours</h1>
+<%= render partial: 'form', locals: { hours: @hours } %>
```
Comment (cblant20) [inline app/views/hours/new.html.erb:2]: here

File: app/views/hours/show.html.erb [hunk 33]
```diff
@@ -0,0 +1,3 @@
+<p><strong><%=  markdown(@hour.title) %></strong><br>
+  <%=  markdown(@hour.body) %>
+</p>
```
Comment (cblant20) [inline app/views/hours/show.html.erb:3]: here

File: app/views/hours/show.html.erb [hunk 34]
```diff
@@ -0,0 +1,3 @@
+<p><strong><%=  markdown(@hour.title) %></strong><br>
```
Comment (cblant20) [inline app/views/hours/show.html.erb:1]: why is hours its own thing?

File: app/views/layouts/application.html.erb [hunk 35]
```diff
@@ -13,17 +13,20 @@
...
-      <li class="nav-li"><%= link_to "FAQ", faq_path %></li>
-      <li class="nav-li"><%= link_to "SUBMISSIONS", submissions_path %></li>
-      <li class="nav-li"><%= link_to "CONTACT", contact_path %></li>
+
+  <i class="ion-navicon-round" data-toggle="collapse" data-target="#mobile"></i>
+  <div id="mobile" class="collapse">
+    <ul class="mobile-menu">
+      <li ><%= link_to "ABOUT", about_path %></li>
+      <li ><%= link_to "ARTISTS", categories_path %></li>
+      <!-- <li ><%= link_to "SHOP" %></li> -->
```
Comment (cblant20) [inline app/views/layouts/application.html.erb:22]: why commented out?

File: app/views/layouts/application.html.erb [hunk 36]
```diff
@@ -36,15 +39,18 @@
         <%= link_to "HOME & SELF CARE", '/categories/6'  %>
       </div>
     </li>
-    <li class="nav-li"><%= link_to "SHOP" %></li>
+    <!-- <li class="nav-li"><%= link_to "SHOP" %></li> -->
```
Comment (cblant20) [inline app/views/layouts/application.html.erb:42]: why is this code duplicateD?

File: app/views/welcome/admin.html.erb [hunk 37]
```diff
@@ -0,0 +1,23 @@
...
+  Look for the Edit and Delete button on each artist profile.</p>
+  <p><strong>Regarding Events</strong><br>
+  Events will automatically be taken of the <%= link_to "new page", new_path %> once their date has passed. To Edit
+  or review old events see the <%= link_to "event index", events_path %>. This can be helpful for recurring events, simply
+  change the date to a future date and it will now show up on the new page.</p>
+  <p><strong>Regarding FAQ</strong><br>
+  New FAQ can be added from this page but to update old FAQ go to the <%= link_to "FAQ Page", faqs_path %> and find the
+  edit and delete buttons for each FAQ in the collapsible panel.</p>
+</div>
+
```
Comment (cblant20) [inline app/views/welcome/admin.html.erb:21]: too many lines down here

File: app/views/welcome/admin.html.erb [hunk 38]
```diff
@@ -0,0 +1,23 @@
+<div class="content-container">
+  <h2 class="category-banner">Admin Panel</h2>
+  <div class="admin-panel">
+    <%= link_to "Update Hours", edit_hour_path(Hour.last), :class => 'btn btn-success' %>
+    <%= link_to "Add Event", new_event_path, :class => 'btn btn-success' %>
+    <%= link_to "View All Events", events_path, :class => 'btn btn-success' %>
+    <%= link_to "Add FAQ", new_faq_path, :class => 'btn btn-success' %>
+  </div>
+  <p><strong>Add Artist</strong><br>
+  To add an artist go to the applicable <%= link_to "category", categories_path %> and use the Add Artist button there</p>
```
Comment (cblant20) [inline app/views/welcome/admin.html.erb:10]: clear instructions here

File: app/views/welcome/admin.html.erb [hunk 39]
```diff
@@ -0,0 +1,23 @@
...
+  <p><strong>To Edit/Delete an artist</strong><br>
+  Look for the Edit and Delete button on each artist profile.</p>
```
Comment (cblant20) [inline app/views/welcome/admin.html.erb:12]: this might be a pain if there are more than 10 artists

File: app/views/welcome/shop.html.erb [hunk 40]
```diff
@@ -0,0 +1,7 @@
+<div class="content-container">
+  <object class='shop-embed' type="text/html" data="https://squareup.com/store/sacred-art">
+    <p>oops! something seems to be wrong with our shopping cart right now, please check back soon.</p>
+  </object>
+</div>
+
```
Comment (cblant20) [inline app/views/welcome/shop.html.erb:6]: extra line

File: db/migrate/20170612161442_create_hours.rb [hunk 41]
```diff
@@ -0,0 +1,10 @@
+class CreateHours < ActiveRecord::Migration
+  def change
+    create_table :hours do |t|
+      t.string :title
+      t.text :body
+
```
Comment (cblant20) [inline db/migrate/20170612161442_create_hours.rb:6]: whatsup with this extra line

File: db/migrate/20170612161912_create_events.rb [hunk 42]
```diff
@@ -0,0 +1,11 @@
+class CreateEvents < ActiveRecord::Migration
+  def change
+    create_table :events do |t|
+      t.string :title
+      t.date :date
+      t.text :body
+
```
Comment (cblant20) [inline db/migrate/20170612161912_create_events.rb:7]: unnecessary line here

File: db/migrate/20170612161958_create_faqs.rb [hunk 43]
```diff
@@ -0,0 +1,10 @@
+class CreateFaqs < ActiveRecord::Migration
+  def change
+    create_table :faqs do |t|
+      t.string :question
+      t.text :answer
+
```
Comment (cblant20) [inline db/migrate/20170612161958_create_faqs.rb:6]: here too

File: db/seeds.rb [hunk 44]
```diff
@@ -40,12 +40,27 @@
...
 
-#csv_text = File.read('Artist_list_category.csv')
-#csv = CSV.parse(csv_text, :headers => true)
-#csv.each do |row|
-#  Artist.create!(row.to_hash)
-#end
+Hour.create!(
+    title:  'Spring Hours',
+    body:   'Monday: By Chance or Appointment
+Tuesday - Friday: noon - 7pm
```
Comment (cblant20) [inline db/seeds.rb:46]: if this is a string amongst multiple lines, you should use a /n

File: db/seeds.rb [hunk 45]
```diff
@@ -40,12 +40,27 @@
...
+  Faq.create!(row.to_hash)
+end
+
+csv_text = File.read('Artist_list_category.csv')
+csv = CSV.parse(csv_text, :headers => true)
+csv.each do |row|
+  Artist.create!(row.to_hash)
+end
 
 puts "#{User.count} users created"
```
Comment (cblant20) [inline db/seeds.rb:62]: don't leave puts in your final product

File: faq.csv [hunk 46]
```diff
@@ -0,0 +1,7 @@
+question,answer
```
Comment (cblant20) [inline faq.csv:1]: interesting way to do this

File: spec/controllers/events_controller_spec.rb [hunk 47]
```diff
@@ -0,0 +1,147 @@
...
+      end
+    end
+
+    describe "EVENT create" do
+      it "returns http success" do
+        expect { post :create, event: { title: "New Awesome Event!", date: "11/11/2017", body: "new awesome things, oh look, they're great!" }}.to change(Event,:count).by(1)
+      end
+
+      it "assigns the new event to @event" do
+        post :create, event: { title: "New Awesome Event!", date: "11/11/2017", body: "new awesome things, oh look, they're great!"}
```
Comment (cblant20) [inline spec/controllers/events_controller_spec.rb:76]: this is repeated, can you hold it in a let?

File: spec/controllers/events_controller_spec.rb [hunk 48]
```diff
@@ -0,0 +1,147 @@
...
+
+        expect(event_instance.id).to eq my_event.id
+        expect(event_instance.title).to eq my_event.title
+        expect(event_instance.body).to eq my_event.body
+        expect(event_instance.date).to eq my_event.date
+      end
+    end
+
+    describe "PUT update" do
+      it "updates artist with expected attributes" do
```
Comment (cblant20) [inline spec/controllers/events_controller_spec.rb:109]: you need to test with null options and missing options

File: spec/controllers/events_controller_spec.rb [hunk 49]
```diff
@@ -0,0 +1,147 @@
...
+        expect(count).to eq 0
+      end
+
+      it "redirects to artist index" do
+        delete :destroy, id: my_event.id
+        expect(response).to redirect_to events_path
+      end
+    end
+  end
+
```
Comment (cblant20) [inline spec/controllers/events_controller_spec.rb:145]: extra lines for no reason?

File: spec/controllers/events_controller_spec.rb [hunk 50]
```diff
@@ -0,0 +1,147 @@
...
+    describe "DELETE destroy" do
+      it "deletes the event" do
+        delete :destroy, id: my_event.id
+        count = Event.where({id: my_event.id}).size
+        expect(count).to eq 0
+      end
+
+      it "redirects to artist index" do
+        delete :destroy, id: my_event.id
+        expect(response).to redirect_to events_path
```
Comment (cblant20) [inline spec/controllers/events_controller_spec.rb:141]: are you testing the flash messages somewhere?

File: spec/controllers/faqs_controller_spec.rb [hunk 51]
```diff
@@ -0,0 +1,166 @@
+require 'rails_helper'
+
+RSpec.describe FaqsController, type: :controller do
+  let(:my_faq) { Faq.create!( question: "What is the answer to life the universe and everything?", answer: "42") }
```
Comment (cblant20) [inline spec/controllers/faqs_controller_spec.rb:4]: you need to test reading from a csv, right?

Comment (cblant20) [inline spec/controllers/faqs_controller_spec.rb:4]: what about a malformed csv? how will you handle it?

File: spec/controllers/faqs_controller_spec.rb [hunk 52]
```diff
@@ -0,0 +1,166 @@
...
+
+    describe "DELETE destroy" do
+      it "returns http redirect" do
+        delete :destroy, id: my_faq.id
+        expect(response).to redirect_to(new_session_path)
+      end
+    end
+  end
+
+  context "admin" do
```
Comment (cblant20) [inline spec/controllers/faqs_controller_spec.rb:59]: you need to have a non-admin test section as well (if it's below, then I am sorry)

File: spec/controllers/hours_controller_spec.rb [hunk 53]
```diff
@@ -0,0 +1,141 @@
...
+      it "returns http success" do
+        get :index
+        expect(response).to have_http_status(:success)
+      end
+
+      it "assigns Hours.all to hour" do
+        get :index
+        expect(assigns(:hours)).to eq([my_hour])
+      end
+    end
```
Comment (cblant20) [inline spec/controllers/hours_controller_spec.rb:62]: you need to test a non-admin and a failing scenario < -for  every case

File: spec/controllers/hours_controller_spec.rb [hunk 54]
```diff
@@ -0,0 +1,141 @@
...
+
+    describe "GET new" do
```
Comment (cblant20) [inline spec/controllers/hours_controller_spec.rb:64]: you also need to test the flash messages everywhere

**Review by cblant20 — CHANGES_REQUESTED** [review]
Here's all the code feedback. got tired at the end. site feedback tomorrow.
